codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoderForBuffer.java

```diff
@@ -0,0 +1,693 @@
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelConfig;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerAdapter;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.socket.ChannelInputShutdownEvent;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.internal.StringUtil;
+
+import java.net.SocketAddress;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * {@link ChannelHandler} which decodes bytes in a stream-like fashion from one {@link Buffer} to an
+ * other Message type.
+ *
+ * For example here is an implementation which reads all readable bytes from
+ * the input {@link Buffer}, creates a new {@link Buffer} and forward it to the next {@link ChannelHandler}
+ * in the {@link ChannelPipeline}.
+ *
+ * <pre>
+ *     public class SquareDecoder extends {@link ByteToMessageDecoderForBuffer} {
+ *         {@code @Override}
+ *         public void decode({@link ChannelHandlerContext} ctx, {@link Buffer} in)
+ *                 throws {@link Exception} {
+ *             ctx.fireChannelRead(in.readBytes(in.readableBytes()));
+ *         }
+ *     }
+ * </pre>
+ *
+ * <h3>Frame detection</h3>
+ * <p>
+ * Generally frame detection should be handled earlier in the pipeline by adding a
+ * {@link DelimiterBasedFrameDecoder}, {@link FixedLengthFrameDecoder}, {@link LengthFieldBasedFrameDecoder},
+ * or {@link LineBasedFrameDecoder}.
+ * <p>
+ * If a custom frame decoder is required, then one needs to be careful when implementing
+ * one with {@link ByteToMessageDecoderForBuffer}. Ensure there are enough bytes in the buffer for a
+ * complete frame by checking {@link Buffer#readableBytes()}. If there are not enough bytes
+ * for a complete frame, return without modifying the reader index to allow more bytes to arrive.
+ * <p>
+ * To check for complete frames without modifying the reader index, use methods like {@link Buffer#getInt(int)}.
+ * One <strong>MUST</strong> use the reader index when using methods like {@link Buffer#getInt(int)}.
+ * For example calling <tt>in.getInt(0)</tt> is assuming the frame starts at the beginning of the buffer, which
+ * is not always the case. Use <tt>in.getInt(in.readerIndex())</tt> instead.
+ * <h3>Pitfalls</h3>
+ * <p>
+ * Be aware that sub-classes of {@link ByteToMessageDecoderForBuffer} <strong>MUST NOT</strong>
+ * annotated with {@link @Sharable}.
+ */
+public abstract class ByteToMessageDecoderForBuffer extends ChannelHandlerAdapter {
+
+    /**
+     * Cumulate {@link Buffer}s by merge them into one {@link Buffer}'s, using memory copies.
+     */
+    public static final Cumulator MERGE_CUMULATOR = (alloc, cumulation, in) -> {
+        // TODO: Is there a replacement for isContiguous()?
```

Comment:
@chrisvest I have added these TODOs here where I was not sure if there is a functionality missing from `Buffer` or I am overlooking something.

Reply:
There's a way to ask if the buffer is composite

Reply:
I'm on phone right now, but if you haven't already there's also a port in the incubator repository you can compare with: https://github.com/netty/netty-incubator-buffer-api/tree/main/buffer-tests/src/test/java/io/netty/buffer/api/tests/examples/bytetomessagedecoder